controller/controller.go

```diff
@@ -112,6 +113,38 @@ type Controller struct {
 	Interval time.Duration
 	// The DomainFilter defines which DNS records to keep or exclude
 	DomainFilter endpoint.DomainFilter
+	// Used by RunOnceThrottled to ensure throttling, once per Interval
+	running      bool
+	minStartTime time.Time
+}
+
+var nowValue = struct{}{}
+
+// Make sure execution happens at most once per interval
+func (c *Controller) RunOnceThrottled(ctx context.Context) error {
+	now := time.Now()
+	testing := false
+	if v, ok := ctx.Value(nowValue).(time.Time); ok {
+		now = v
+		testing = true
+	}
+	if c.running || now.Before(c.minStartTime) {
+		// Normally when throttled this function is no-op, but when testing
+		// we need to signal that the function is throttled by returning an error
```

Comment:
Can you rephrase this to be easier to understand? I'm not sure I understand what you mean with ` Normally when throttled this function is no-op`. What you are signaling here is exactly that you are throttled.. or I misunderstand 😄

Reply:
The key is "but when testing" :) Normally it's no-op, but when testing this function returns an error when it's throttled so we can make assertion in tests.